minio/minio
# Add support for bucket replication

## Description
[S3](https://docs.aws.amazon.com/AmazonS3/latest/dev/replication-how-setup.html) has a bucket replication API. This PR adds support for a similar API within MinIO

## Motivation and Context
Allows synchronous mirroring

## How to test this PR?
Note: This needs [minio-go PR](https://github.com/minio/minio-go/pull/1314) and [mc PR](https://github.com/minio/mc/pull/3272)

Since this PR has MinIO  specific tweak to use replication ARN to identify the replica endpoint, you will not be able to use aws cli to set the replication configuration as replication ARN needs to be validated..
```
#!/bin/sh
mc mb myminio/source --with-lock
mc mb myminio/dest  --with-lock
aws s3api put-bucket-versioning --bucket source --versioning-configuration Status=Enabled --endpoint-url http://localhost:9000 --profile minio
aws s3api put-bucket-versioning --bucket dest --versioning-configuration Status=Enabled --endpoint-url http://localhost:9000 --profile minio
mc admin bucket replication set myminio/source http://minio:minio123@localhost:9000/dest --path-style "auto" --api "s3v2"  # to associate replication endpoint & target bucket to source bucket
mc admin bucket remote myminio http://localhost:9000  # to fetch replication ARN for target endpoint
mc replicate myminio/source ARN--config /tmp/replication-config.json
```
minio-source-replication-config.json
```
{
  "Role": "arn:aws:iam::xxx:role/replicationRole",
  "Rules": [
    {
      "Status": "Enabled",
      "Priority": 1,
      "DeleteMarkerReplication": { "Status": "Disabled" },
      "Filter" : { "Prefix": "Backups"},
      "Destination": {
        "Bucket": "arn:aws:s3:::dest",
        "StorageClass":"STANDARD"
      }
    }
  ]
}
```

Do a `mc cp` to source with the filter criteria being met on prefix and tags if any. You should see the `X-Amz-Replication-Status` on source and dest reflecting the status of the replication
## Types of changes

## Checklist:
